draft-pda-core.md

```diff
@@ -291,11 +291,437 @@ Each PA task in this document is divided into three sub-protocols as follows.
+
+Each helper specifies the
+[HPKE](https://datatracker.ietf.org/doc/draft-irtf-cfrg-hpke/) public key the
+client will use to encrypt the helper's share of the report. The public key and
+associated parameters are structured as follows:
+
+~~~
+struct {
+  uint8 id;
+  HpkeKemId kem_id;
+  HpkeKdfId kdf_id;
+  HpkeAeadKdfId aead_id;
+  HpkePublicKey public_key;
+} HPKEKeyConfig;
+
+opaque HpkePublicKey<0..2^16-1>;
+uint16 HpkeAeadId; // Defined in I-D.irtf-cfrg-hpke
+uint16 HpkeKemId;  // Defined in I-D.irtf-cfrg-hpke
+uint16 HpkeKdfId;  // Defined in I-D.irtf-cfrg-hpke
+~~~
+
+We call this the helper's *key configation*. The key configuration is used to
+set up a base-mode HPKE context to use to derive symmetric keys for protecting
+the shares sent to the helper. The *config id*, `HPKEKeyConfig.id`, is forwarded
+by the client to the helper, who uses this value to decide if it knows how to
+decrypt a share it receives.
+
+**Pre-conditions.**
+We assume the following conditions hold before the protocol begins:
+1. The client, aggregators, and collector are configured with a specific PA task.
+1. The client knows the URL of the leader endpoint, e.g., `example.com/metrics`.
+   We write this URL as `[leader]` below. (We write `[helper]` for a helper's
+   URL.)
+1. The client and leader can establish a leader-authenticated TLS channel.
+1. The leader and each helper can establish a leader-authenticated TLS channel.
+1. Each helper has chosen an HPKE key pair.
+1. The aggregators agree on a set of PA tasks, as well as the PA protocol and
+   parameters used for each task.
+
+## Upload {#pa-upload}
+
+[TODO: Add an illustration of this sub-protocol.]
+
+Uploading a report involves two requests to the leader. In the  *upload start
+request*, the client discovers the protocol-specific parameters and the endpoint
+URL of each helper. In the *upload finish request* uploads its report to the
+leader.
+
+[OPEN ISSUE: @acmiyaguchi pointed out that the use of an anonymizing proxy for
+uploading shares might be easier to implement if the "upload" phase involved a
+single HTTP request. Is this true? Does OHTTP
+(https//www.ietf.org/archive/id/draft-thomson-http-oblivious-01.html) allow
+clients to make multiple requests without too much fuss?]
+
+### Upload Start
+
+**Request.**
+Let `[leader]` denote the URL of the leader's report ingestion endpoint. The
+client sends a POST request to `[leader]/upload_start` with the following
+message:
+
+~~~
+struct {
+  PATask task;
+  opaque id[16];
+} PAUploadStartReq;
+~~~
+
+The first field, `task` corresponds to the PA task for which a report will be
+generated. The second, `id` is a random, 16-byte string generated by the
+client.
+
+[OPEN ISSUE: Maybe use a UUID for `PAUploadSTartReq.id`, as well as for
+`PATask.id`?]
+
+**Response.**
+The leader responds to well-formed requests to `[leader]/upload_start` with
+status 200 and the following message:
+
+~~~
+struct {
+  Url helper_urls<0..2^16-1>;
```

Comment:
```suggestion
  Url helper_urls<1..2^16-1>;
```

Also, has `Url` been defined yet?

Reply:
It hasn't! Added a definition.

I'm not taking the suggestion because I don't want to get nit-picky about size ranges yet. This is something that's likely to change a lot.

Reply:
Well, can the helper_urls be empty? If not, then we should take the suggestion.

Reply:
Well, a Url is itself a u16-prefixed string, so I guess the minimum is actually 2. Again, I don't think it's worth being picky about this right now.

Reply:
It's not really being picky. It's just saying that the URLs must be non-empty.